Initium-Odp/src/com/universeprojects/miniup/server/commands/CommandAttack.java

```diff
@@ -0,0 +1,200 @@
+import java.util.GregorianCalendar;
+import java.util.Map;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.universeprojects.cacheddatastore.CachedDatastoreService;
+import com.universeprojects.cacheddatastore.CachedEntity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key;
+import com.universeprojects.miniup.server.GameUtils;
+import com.universeprojects.miniup.server.ODPDBAccess;
+import com.universeprojects.miniup.server.commands.framework.Command;
+import com.universeprojects.miniup.server.commands.framework.UserErrorMessage;
+
+
+/**
+ * Allows the player to attack a mob at the current location
+ * 
+ * Usage notes:
+ * Checks if caller can attack and initiates combat if so
+ * 
+ * Parameters:
+ * 		charId - characterId of monster to attack 
+ * 
+ * Support methods:
+ * 		canAttack - returns boolean whether attack is allowed
+ * 
+ * @author NJ
+ *
+ */
+public class CommandAttack extends Command {
+
+	public CommandAttack(HttpServletRequest request, HttpServletResponse response) 
+	{
+		super(request, response);
+	}
+
+	public static boolean canAttack(CachedDatastoreService ds, CachedEntity location, CachedEntity character, CachedEntity monster)
+	{
+		try {
+			canAttackHelper(ds, location, character, monster);
+			return true;
+		} catch (Exception e) {
+			return false;
+		}
+	}
+	
+	private static void canAttackHelper(CachedDatastoreService ds, CachedEntity location, CachedEntity character, CachedEntity monster) throws UserErrorMessage
+	{
+		if (ds==null || location==null || character==null || monster==null)
+			throw new RuntimeException("canAttackHelper invalid parameter.");
+		
+		// Check that monster is an actual monster
+		if ("NPC".equals(monster.getProperty("type"))==false)
+			throw new UserErrorMessage("You can only attack monsters this way.");
+		
+		// Check that they're both at the same location
+		if (GameUtils.equals(location.getKey(), character.getProperty("locationKey"))==false || GameUtils.equals(location.getKey(), monster.getProperty("locationKey"))==false)
+			throw new UserErrorMessage("You can only attack monsters in the immediate vicinity.");
+		
+		// Check that character status is set to Normal (or blank)
+		String mode = (String)character.getProperty("mode");
+		if (mode!=null && mode.equals("NORMAL")==false)
+			throw new UserErrorMessage("You're too busy to attack anything at the moment.");
+		
+		// Additional incapacitated check just in case mode wasn't set properly.
+		if (GameUtils.isPlayerIncapacitated(character))
+			throw new UserErrorMessage("You're incapacitated at the moment.");
+		
+		// Check if monster is alive
+		Double hitpoints = (Double)monster.getProperty("hitpoints");
+		mode = (String)monster.getProperty("mode");
+		if (hitpoints==null || hitpoints<=0 || "DEAD".equals(mode))
+		{
+			// DB sanity check. Mode should be DEAD
+			if ("DEAD".equals(mode)==false)
```

Comment:
You know what... maybe instead of having this sort of code all over the place we should have a "boolean GameUtils.normalizeDatabaseState(CachedEntity entity)" call which would do all sorts of different checks to make sure everything makes sense.